File: morning_exercises/day4.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

std::vector<std::vector<int>> twoSum(std::vector<int>& nums, int target) {
    std::vector<std::vector<int>> res;
    std::unordered_map<int, int> seen;

    if(nums.size()==0) return res;

    for (int i=0;i<nums.size();i++) {
        const int complement = target - nums [i];

        if(seen.count(complement)) res.push_back({seen.at(complement), i});

        seen[nums[i]]=i;
    }
    return res;
}

int main() {
    std::vector<int> nums = {1,2,3,4};

    auto res = twoSum(nums, 5);
    for (const auto& pair : res) std::cout << "[" << pair[0] << "," << pair[1] << "]\n";
}

File: morning_exercises/day4.ts
function twoSum(nums: number[], target: number): number[][] {
    let res: number[][]=[];
    if (nums.length===0) return []

    for(let i = 0; i<nums.length; i++) {
        for(let j = i+1; j<nums.length; j++) {
            if ((nums[i]+nums[j])==target) res.push([i,j]);
        }
    }
    return res
}

// console.log(twoSum([1,2,3,4], 5));

function twoSumOptimal(nums: number[], target: number): number[][] {
    let res:number[][]=[];
    const seen = new Map<number, number>();

    if ((nums.length)===0) return [];

    for (let i=0;i<nums.length; i++) {
        const compliment = target - nums[i]!;

        if (seen.has(compliment)) res.push([(seen.get(compliment)!),i]);
        seen.set(nums[i]!,i);
    }
    return res;
}

console.log(JSON.stringify(twoSumOptimal([1,2,3,4], 5)));